File: app/src/main/java/miles/scribble/home/drawing/BitmapDrawHistory.java
package miles.scribble.home.drawing;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.net.Uri;

import java.io.IOException;
import java.io.InputStream;

/**
 * Created by mbpeele on 1/4/16.
 */
class BitmapDrawHistory {

    public Uri uri;
    public float[] matrixValues;

    public BitmapDrawHistory(Uri uri, float[] matrixValues) {
        this.uri = uri;
        this.matrixValues = matrixValues;
    }

    public void draw(Matrix matrix, BitmapCache cache, Context context, Canvas canvas) {
        float[] prevMatrixValues = new float[9];
        matrix.getValues(prevMatrixValues);

        Bitmap bitmap = cache.get(uri);
        if (bitmap == null) {
            InputStream inputStream;
            try {
                inputStream = context.getContentResolver().openInputStream(uri);
                bitmap = cache.decodeStream(inputStream);

                if (inputStream != null) {
                    inputStream.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        matrix.setValues(matrixValues);

        canvas.save();
        canvas.concat(matrix);
        canvas.drawBitmap(bitmap, 0, 0, null);
        canvas.restore();

        matrix.setValues(prevMatrixValues);
    }
}

File: app/src/main/java/miles/scribble/home/drawing/drawhistory/BitmapDrawHistory.kt
package miles.scribble.home.drawing.drawhistory

import android.content.Context
import android.graphics.Canvas
import android.graphics.Matrix
import android.net.Uri
import miles.scribble.util.BitmapCache

import java.io.IOException
import java.io.InputStream

/**
 * Created by mbpeele on 1/4/16.
 */
internal class BitmapDrawHistory(var uri: Uri, var matrixValues: FloatArray) {

    fun draw(matrix: Matrix, cache: BitmapCache, context: Context, canvas: Canvas) {
        val prevMatrixValues = FloatArray(9)
        matrix.getValues(prevMatrixValues)

        var bitmap = cache.get(uri)
        if (bitmap == null) {
            val inputStream: InputStream?
            try {
                inputStream = context.contentResolver.openInputStream(uri)
                bitmap = cache.decodeStream(inputStream)

                inputStream?.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }

        matrix.setValues(matrixValues)

        canvas.save()
        canvas.concat(matrix)
        canvas.drawBitmap(bitmap, 0f, 0f, null)
        canvas.restore()

        matrix.setValues(prevMatrixValues)
    }
}
